libs/vkd3d/swapchain.c

```diff
@@ -0,0 +1,2252 @@
+{
+    struct d3d12_swapchain_factory *swapchain_factory = d3d12_swapchain_factory_from_IWineDXGISwapChainFactory(iface);
+    ULONG refcount = InterlockedIncrement(&swapchain_factory->refcount);
+
+    TRACE("%p increasing refcount to %u.\n", swapchain_factory, refcount);
+
+    return refcount;
+}
+
+static ULONG STDMETHODCALLTYPE d3d12_swapchain_factory_Release(IWineDXGISwapChainFactory *iface)
+{
+    struct d3d12_swapchain_factory *swapchain_factory = d3d12_swapchain_factory_from_IWineDXGISwapChainFactory(iface);
+    ULONG refcount = InterlockedDecrement(&swapchain_factory->refcount);
+
+    /* Ignore refcount for this...
+     * This gets freed in the device at cleanup time... */
+
+    TRACE("%p decreasing refcount to %u.\n", swapchain_factory, refcount);
+
+    return refcount;
+}
+
+static HRESULT STDMETHODCALLTYPE d3d12_swapchain_factory_QueryInterface(IWineDXGISwapChainFactory *iface,
+        REFIID iid, void **out)
+{
+    TRACE("iface %p, iid %s, out %p.\n", iface, debugstr_guid(iid), out);
+
+    if (IsEqualGUID(iid, &IID_IWineDXGISwapChainFactory) ||
+        IsEqualGUID(iid, &IID_IUnknown))
+    {
+        d3d12_swapchain_factory_AddRef(iface);
+        *out = iface;
+        return S_OK;
+    }
+
+    WARN("%s not implemented, returning E_NOINTERFACE.\n", debugstr_guid(iid));
+
+    *out = NULL;
+    return E_NOINTERFACE;
+}
+
+static HRESULT STDMETHODCALLTYPE d3d12_swapchain_factory_CreateSwapChainForHwnd(IWineDXGISwapChainFactory *iface,
+        IDXGIFactory* factory, HWND window, const DXGI_SWAP_CHAIN_DESC1* swapchain_desc,
+        const DXGI_SWAP_CHAIN_FULLSCREEN_DESC* fullscreen_desc, IDXGIOutput* output,
+        IDXGISwapChain1** swapchain)
+{
+    struct d3d12_swapchain_factory *swapchain_factory = d3d12_swapchain_factory_from_IWineDXGISwapChainFactory(iface);
+    HRESULT hr;
+
+    TRACE("iface %p, factory %p, window %p, swapchain_desc %p, fullscreen_desc %p, output %p, swapchain %p\n",
+        iface, factory, window, swapchain_desc, fullscreen_desc, output, swapchain);
+
+    hr = d3d12_swapchain_create(factory, swapchain_factory->queue, window, swapchain_desc, fullscreen_desc, swapchain);
+
+    return hr;
+}
+
+static const struct IWineDXGISwapChainFactoryVtbl d3d12_swapchain_factory_vtbl =
+{
+    /* IUnknown methods */
+    d3d12_swapchain_factory_QueryInterface,
+    d3d12_swapchain_factory_AddRef,
+    d3d12_swapchain_factory_Release,
+
+    /* IWineDXGISwapChainFactory methods */
+    d3d12_swapchain_factory_CreateSwapChainForHwnd
+};
+
+static HRESULT d3d12_swapchain_factory_init(struct d3d12_command_queue *queue, struct d3d12_swapchain_factory *factory)
+{
+    factory->IWineDXGISwapChainFactory_iface.lpVtbl = &d3d12_swapchain_factory_vtbl;
+    factory->refcount = 1;
+    factory->queue = queue;
+
+    return S_OK;
+}
+
+HRESULT d3d12_swapchain_factory_create(struct d3d12_command_queue *queue, IWineDXGISwapChainFactory** factory)
```

Comment:
This whole thing is **way** overengineerd. Do you expect to add anything that warrants a memory allocation? Can't we just share the ref count with the queue?

I'd much rather have just the init function and nothing else, and make the struct itself a member of the queue rather than a pointer to it, this would also allow us to get rid of half the `#ifdefs` in common code.

Reply:
Resolved locally.